core/src/services/aliyun_drive/writer.rs

```diff
@@ -44,162 +43,89 @@ impl AliyunDriveWriter {
-                    }
+            _ => {
+                let res = self
+                    .core
+                    .create(
+                        Some(&self.parent_file_id),
+                        &self.name,
+                        CreateType::File,
+                        CheckNameMode::Refuse,
+                    )
+                    .await?;
+                let output: CreateResponse =
+                    serde_json::from_reader(res.reader()).map_err(new_json_deserialize_error)?;
+                if output.exist.is_some_and(|x| x) {
+                    return Err(Error::new(ErrorKind::AlreadyExists, "file exists"));
                 }
+                let upload_id = output.upload_id.expect("cannot find upload_id");
+                let file_id = output.file_id;
+                self.upload_id = Some(upload_id.clone());
+                self.file_id = Some(file_id.clone());
+                (upload_id, file_id)
             }
-        }
-
-        Ok(output.upload_id)
-    }
-
-    async fn complete(&self, upload_id: &str) -> Result<Buffer> {
-        let file_id = self.read_file_id().await?;
-
-        self.core.complete(&file_id, upload_id).await
-    }
-
-    async fn delete(&self) -> Result<()> {
-        let file_id = self.read_file_id().await?;
+        };
 
-        self.core.delete_path(&file_id).await
+        Ok((upload_id, file_id))
     }
 }
 
-impl oio::MultipartWrite for AliyunDriveWriter {
-    async fn write_once(&self, size: u64, body: crate::Buffer) -> Result<()> {
-        let upload_id = self.initiate_part().await?;
-        self.write_part(&upload_id, 0, size, body).await?;
-
-        self.complete(&upload_id).await?;
-        Ok(())
-    }
-
-    async fn initiate_part(&self) -> Result<String> {
-        let Some(upload_id) = self.write(None, None).await? else {
-            return Err(Error::new(ErrorKind::Unsupported, "cannot find upload_id"));
-        };
-
-        Ok(upload_id)
-    }
+impl oio::Write for AliyunDriveWriter {
+    async fn write(&mut self, bs: Buffer) -> Result<usize> {
+        let (upload_id, file_id) = self.get_upload_info(false).await?;
 
-    async fn write_part(
-        &self,
-        upload_id: &str,
-        part_number: usize,
-        _size: u64,
-        body: crate::Buffer,
-    ) -> Result<oio::MultipartPart> {
-        let file_id = self.read_file_id().await?;
         let res = self
             .core
-            .get_upload_url(&file_id, upload_id, Some(part_number + 1))
+            .get_upload_url(&file_id, &upload_id, Some(self.part_number))
             .await?;
         let output: UploadUrlResponse =
             serde_json::from_reader(res.reader()).map_err(new_json_deserialize_error)?;
 
-        let Some(part_info_list) = output.part_info_list else {
-            return Err(Error::new(
-                ErrorKind::Unexpected,
-                "cannot find part_info_list",
-            ));
+        let upload_url = match output.part_info_list {
```

Comment:
We can use `let Some(part_info_list) = output.part_info_list else {}`

Reply:
make sense